crazy_rl/multi_agent/jax/surround/surround.py

```diff
@@ -0,0 +1,203 @@
+from typing_extensions import override
+
+import jax.numpy as jnp
+import jax_dataclasses as jdc
+import numpy as np
+from gymnasium import spaces
+from jax import jit, vmap
+
+from crazy_rl.multi_agent.jax.base_parallel_env import BaseParallelEnv
+
+
+@override
+@jdc.pytree_dataclass
+class State:
+    """State of the environment containing the modifiable variables."""
+
+    agent_location: jnp.ndarray
+    timestep: int
+    terminations: jnp.ndarray = jnp.array([])
+    rewards: jnp.ndarray = jnp.array([])
+    observations: jnp.ndarray = jnp.array([])
+    truncations: jnp.ndarray = jnp.array([])
+
+
+class Surround(BaseParallelEnv):
+    """A Parallel Environment where drone learn how to surround a target point."""
+
+    metadata = {"render_modes": ["human", "real"], "is_parallelizable": True, "render_fps": 20}
+
+    def __init__(
+        self,
+        num_drones: int,
+        init_flying_pos: jnp.ndarray,
+        target_location: jnp.ndarray,
+        render_mode=None,
+        size: int = 4,
+        swarm=None,
+    ):
+        """Surround environment for Crazyflies 2.
+
+        Args:
+            num_drones: Number of drones
+            init_flying_pos: Array of initial positions of the drones when they are flying
+            target_location: Array of the position of the target point
+            render_mode: Render mode: "human", "real" or None
+            size: Size of the map
+            swarm: Swarm object, used for real tests. Ignored otherwise.
+        """
+        self.state = State(jnp.copy(init_flying_pos), 0)
+
+        # Constant variables
+        self.num_drones = num_drones
+
+        self._target_location = target_location  # unique target location for all agents
+
+        self._init_flying_pos = init_flying_pos
+
+        self.size = size
+
+        self.norm = vmap(jnp.linalg.norm)  # function to compute the norm of each array in a matrix
+
+        super().__init__(
+            render_mode=render_mode,
+            size=size,
+            init_flying_pos=self._init_flying_pos,
+            target_location=self._target_location,
+            num_drones=self.num_drones,
+            swarm=swarm,
+            state=self.state,
+        )
+
+    @override
+    def _observation_space(self, agent):
+        return spaces.Box(
+            low=np.tile(np.array([-self.size, -self.size, 0], dtype=np.float32), self.num_drones + 1),
+            high=np.tile(np.array([self.size, self.size, self.size], dtype=np.float32), self.num_drones + 1),
+            shape=(3 * (self.num_drones + 1),),  # coordinates of the drones and the target
+            dtype=jnp.float32,
+        )
+
+    @override
+    def _action_space(self):
+        return spaces.Box(low=-1 * np.ones(3, dtype=np.float32), high=np.ones(3, dtype=np.float32), dtype=np.float32)
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_obs(self, state):
+        obs = jnp.array([[] for _ in range(self.num_drones)])
```

Comment:
I think something like that gives a matrix already: `jnp.tile(state.agents_locations, (self.num_drones, 1))`

What I would do to vectorize is probably:
1. copy state.agents_locations
2. append target_location
3. generate a full matrix using the code above
4. find a way to swap elements in the array like using slices: https://www.w3resource.com/python-exercises/numpy/python-numpy-exercise-150.php